Improve error message on invalid channel config after init timeout

@davidrohr wrote on Nov 16, 2022:
> we had runs failing to start with an Error:
> `fairmq/Device.cxx:            LOG(error) << "could not connect all channels after " << fInitializationTimeoutInS << " attempts";` which comes from FairMQ directly.
> [...] could you add additional output printing which channel(s) failed to connect?

There is actually already some hints as to which channels are still invalid in debug logs (although it can be quite spammy for many channels and long init times):
```
[19:51:18][DEBUG] Validating channel 'data[0]'... INVALID
[19:51:18][DEBUG] invalid channel address: 'unspecified'
[19:51:18][DEBUG] Validating channel 'data[0]'... INVALID
[19:51:18][DEBUG] invalid channel address: 'unspecified'
[19:51:18][ERROR] could not connect all channels after 120 attempts
```

But I agree that it makes sense to list the channels also with the error. I extend the error to:
```

[19:51:18][ERROR] could not connect all channels after 120 attempts
[19:51:18][ERROR] following channels are still invalid:
[19:51:18][ERROR] channel: name: data[0], type: push, method: connect, address: unspecified
[19:51:18][DEBUG] Exception caught in ProcessWork(), going to Error state
```
in #455.

--- a/fairmq/Channel.h
+++ b/fairmq/Channel.h
@@ -19,6 +19,7 @@
 
 #include <cstdint>   // int64_t
 #include <memory>   // unique_ptr, shared_ptr
+#include <ostream>
 #include <stdexcept>
 #include <string>
 #include <utility>   // std::move
@@ -379,6 +380,14 @@ class Channel
     static constexpr int DefaultPortRangeMin = 22000;
     static constexpr int DefaultPortRangeMax = 23000;
     static constexpr bool DefaultAutoBind = true;
+
+    friend std::ostream& operator<<(std::ostream& os, const Channel& ch)
+    {
+        return os << "name: " << ch.fName
+                  << ", type: "  << ch.fType
+                  << ", method: "  << ch.fMethod
+                  << ", address: "  << ch.fAddress;
+    }
 
   private:
     std::shared_ptr<TransportFactory> fTransportFactory;

--- a/fairmq/Device.cxx
+++ b/fairmq/Device.cxx
@@ -304,7 +304,7 @@ void Device::ConnectWrapper()
     // first attempt
     AttachChannels(fUninitializedConnectingChannels);
     // if not all channels could be connected, update their address values from config and retry
-    while (!fUninitializedConnectingChannels.empty()) {
+    while (!fUninitializedConnectingChannels.empty() && !NewStatePending()) {
         this_thread::sleep_for(chrono::milliseconds(sleepTimeInMS));
 
         for (auto& chan : fUninitializedConnectingChannels) {

--- a/fairmq/Device.cxx
+++ b/fairmq/Device.cxx
@@ -317,6 +317,10 @@ void Device::ConnectWrapper()
 
         if (numAttempts++ > maxAttempts) {
             LOG(error) << "could not connect all channels after " << fInitializationTimeoutInS << " attempts";
+            LOG(error) << "following channels are still invalid:";
+            for (auto& chan : fUninitializedConnectingChannels) {
+                LOG(error) << "channel: " << *chan;
+            }
             throw runtime_error(tools::ToString("could not connect all channels after ", fInitializationTimeoutInS, " attempts"));
         }
 

